onnxruntime/core/providers/vitisai/vitisai_execution_provider.cc

```diff
@@ -1,300 +1,145 @@
+#include "core/session/custom_ops.h"
+#include "core/session/inference_session.h"
+
+#include "onnxruntime_vitisai_ep/onnxruntime_vitisai_ep.h"
 
 using namespace ONNX_NAMESPACE;
-using namespace ::onnxruntime::logging;
 
 namespace onnxruntime {
 
 constexpr const char* VITISAI = "VITISAI";
 
-typedef std::shared_ptr<pyxir::graph::XGraph> XGraphHolder;
-typedef std::shared_ptr<pyxir::graph::XLayer> XLayerHolder;
-
-VitisAIExecutionProvider::VitisAIExecutionProvider(const VitisAIExecutionProviderInfo& info)
-    : IExecutionProvider{onnxruntime::kVitisAIExecutionProvider}, backend_type_(info.backend_type), device_id_(info.device_id), export_runtime_module_(info.export_runtime_module), load_runtime_module_(info.load_runtime_module) {
-  AllocatorCreationInfo default_memory_info{
-      [](int) {
-        return std::make_unique<CPUAllocator>(OrtMemoryInfo(VITISAI, OrtAllocatorType::OrtDeviceAllocator));
-      }};
-
-  InsertAllocator(CreateAllocator(default_memory_info));
+static vaip_core::DllSafe<std::vector<std::unique_ptr<vaip_core::ExecutionProvider>>> compile_onnx_model(
+    const onnxruntime::GraphViewer& graph_viewer,
+    const logging::Logger& logger, const char* json_config) {
+#ifndef _WIN32
+  auto model_path = graph_viewer.ModelPath().ToPathString();
+#else
+  using convert_t = std::codecvt_utf8<wchar_t>;
+  std::wstring_convert<convert_t, wchar_t> strconverter;
+  auto model_path = strconverter.to_bytes(graph_viewer.ModelPath().ToPathString());
+#endif
+  return onnxruntime_vitisai_ep::compile_onnx_model_3(model_path, graph_viewer.GetGraph(), json_config);
 }
-
-/**
- * Returns a vector of clusters (or node_idx) that are supported by the given
- * backend type
- */
-static std::vector<std::vector<NodeIndex>>
-GetSupportedNodeClusters(const XGraphHolder& xg, const std::string& backend_type,
-                         const GraphViewer& graph_viewer,
-                         /*out*/ std::unordered_set<std::string>& required_initializers) {
-  std::vector<std::vector<NodeIndex>> clusters;
-
-  // Retrieve supported tensor names and corresponding subgraphs they belong to
-  int cur_idx = 0;
-  std::unordered_map<std::string, std::string> supported_tensors;
-  std::unordered_map<std::string, int> cluster_idx;
-  for (auto& xl_name : xg->get_layer_names()) {
-    XLayerHolder xl = xg->get(xl_name);
-    if (xl->target == backend_type) {
-      supported_tensors[xl->get_attr("onnx_id").get_string()] = xl->subgraph;
-      if (cluster_idx.find(xl->subgraph) == cluster_idx.end()) {
-        cluster_idx[xl->subgraph] = cur_idx;
-        std::vector<NodeIndex> new_cluster;
-        clusters.push_back(new_cluster);
-      }
-    }
+struct MyCustomOpKernel : OpKernel {
+  MyCustomOpKernel(const OpKernelInfo& info, const OrtCustomOp& op) : OpKernel(info), op_(op) {
+    op_kernel_ = op_.CreateKernel(&op_, OrtGetApiBase()->GetApi(op_.version),
+                                  reinterpret_cast<const OrtKernelInfo*>(&info));
```

Comment:
What's the reason for using the external ORT API for internal EP code?

Reply:
In the future, we will put the VitisAI EP into an independent DLL which can be separately compiled.

We don't know whether there are alternative internal ORT APIs. Do you have any suggestion so that we can update accordingly.